### PR TITLE
On check_timeout first send SIGTERM

### LINKED ISSUE #6162: Check timeouts: gracefully terminate check by sending SIGTERM before SIGKILL
When a check command timeouts, Icinga2 currently sends it a SIGKILL, not allowing it to shut down gracefully. It should first send a SIGTERM and only if the process then doesn't terminate within a few seconds send a SIGKILL. This would for example allow checks to remove temporary files they created, which isn't really possible at the moment without implementing a separate timeout mechanism in the check and keeping the timeout values in sync so that the check's timeout fires a few seconds before Icinga would kill it.

This was also suggested on the icinga-users mailinglist a while ago but got no response there: https://lists.icinga.org/pipermail/icinga-users/2016-February/010822.html

### PR DESCRIPTION
**Pull request by Al2Klimov**
... to allow check plugins to terminate gracefully.

fixes #6162

### CODE REVIEW

**Comment by Al2Klimov** [comment]
```
[2020-03-17 15:53:58 +0100] warning/Process: Terminating process 21176 ('/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_lolcat') after timeout of 10 seconds
[2020-03-17 15:53:59 +0100] warning/Process: Killing process group 21176 ('/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_lolcat') after timeout of 11 seconds
[2020-03-17 15:53:59 +0100] warning/Process: PID 21176 was terminated by signal 9 (Killed: 9)
[2020-03-17 15:53:59 +0100] warning/PluginCheckTask: Check command for object 'Alexanders-MacBook-Pro.local!lolcat' (PID: 21176, arguments: '/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_lolcat') terminated with exit code 128, output: <Timeout exceeded.><Terminated by signal 9 (Killed: 9).>
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
If I understand correctly (haven't yet actually run the code), once the timeout is exceeded, this will now set `timeout` to a value less than 0: https://github.com/Icinga/icinga2/blob/85993b7ff9dfb4a2fb50cd99fdfc414f35805565/lib/base/process.cpp#L664-L667

Therefore `poll()` will now always wait for the minimum duration of 500ms while we're waiting: https://github.com/Icinga/icinga2/blob/85993b7ff9dfb4a2fb50cd99fdfc414f35805565/lib/base/process.cpp#L674-L682

I guess we should think a bit if we can find a nicer way to do this without having to rewrite too much code, or if we just accept these 500ms intervals (shouldn't really be a problem).

**Comment by Al2Klimov** [comment]
That's another construction area, isn't it?

File: lib/base/process.cpp [hunk 1]
```diff
@@ -50,10 +50,13 @@ static boost::once_flag l_SpawnHelperOnceFlag = BOOST_ONCE_INIT;
...
 Process::Process(Process::Arguments arguments, Dictionary::Ptr extraEnvironment)
 	: m_Arguments(std::move(arguments)), m_ExtraEnvironment(std::move(extraEnvironment)),
-	  m_Timeout(600), m_AdjustPriority(false), m_ResultAvailable(false)
+	  m_Timeout(600)
 #ifdef _WIN32
 	, m_ReadPending(false), m_ReadFailed(false), m_Overlapped()
+#else /* _WIN32 */
+	, m_Terminated(false)
 #endif /* _WIN32 */
+	, m_AdjustPriority(false), m_ResultAvailable(false)
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/base/process.cpp:59]: To order initializations as the definitions.

**Comment by julianbrost** [comment]
> That's another construction area, isn't it?

I guess you are referring to the 500ms polling intervals: I think this is newly introduced by this PR. The process is removed from the poll list when `DoEvents()` returns `false`: https://github.com/Icinga/icinga2/blob/5f548c8f8965eeeacfcd218239a1f89bde3e30c9/lib/base/process.cpp#L730-L739

And the only places where `true` is returned, are within `if (!is_timeout)`: https://github.com/Icinga/icinga2/blob/5f548c8f8965eeeacfcd218239a1f89bde3e30c9/lib/base/process.cpp#L1056-L1081

So if I'm reading the code correctly, until now, `DoEvents()` is done only once after a timeout is reached and the process is removed from the poll list. This PR introduces a new code path for sending SIGTERM that happens after the timeout but does not set `is_timeout = true`. Therefore, the process may now remain in the poll list (and according to your tests does, as you have entered `DoEvents()` a second time where a SIGKILL is sent. And this polling whether SIGKILL is to be sent now happens every 500ms, regardless of the actual them when it is supposed to be sent.

**Comment by Al2Klimov** [comment]
Duuude... if I'd not oversee 🐘s all the time, I'd say you've Xray in you Stark Industries glasses. 😉

File: lib/base/process.cpp [hunk 2]
```diff
@@ -716,7 +719,11 @@ void Process::IOThreadProc(int tid)
 				bool is_timeout = false;
 
 				if (it->second->m_Timeout != 0) {
-					double timeout = it->second->m_Result.ExecutionStart + it->second->m_Timeout;
+					double timeout = it->second->m_Result.ExecutionStart + it->second->m_Timeout
+#ifndef _WIN32
+						* (it->second->m_SentSigterm ? 1.1 : 1.0)
+#endif /* _WIN32 */
+						;
```
Comment (julianbrost) [inline lib/base/process.cpp:726]: I think just doing the following would be nicer, but also see the general comment about adding a new method for this.

```cpp
#ifndef _WIN32
    if (it->second->m_SentSigterm) {
        timeout *= 1.1;
    }
#endif
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
The correct timeout should also be used there: https://github.com/Icinga/icinga2/blob/de2d117a04deb519183bf9c982f6a4ea9bf08984/lib/base/process.cpp#L663-L668

In general, it would probably be a good idea to add something like a `GetNextTimeout()` method, that returns either `m_Timeout` or `m_Timeout * 1.1`. This would also avoid having `1.1` scattered around as a magic constant.

**Review by julianbrost — APPROVED** [review]
LGTM in general, please check the one comment in the code, maybe you have a good idea there. Tested the following:

```
object CheckCommand "sigterm" {
	command = [ConfigDir + "/scripts/check_sigterm", "$sigterm_sleep$"]
	vars.sigterm_sleep = 0
}

apply Service "sigerm-" for (d in range(10)) {
	assign where host.name == "master-1"
	import "generic-service"
	command_endpoint = host.name
	check_command = "sigterm"
	check_interval = 120
	retry_interval = 120
	check_timeout = 60
	vars.sigterm_sleep = d
}
```

Where this is `check_sigterm` (i.e. exits `$1` seconds after receiving SIGTERM):
```bash
#!/bin/bash

set -eu

d=$1

sigterm_handler() {
	printf 'Received SIGTERM. '
	sleep $d
	printf 'Exiting. '
	exit 0
}

trap sigterm_handler SIGTERM

printf 'Waiting. '
sleep inf &
wait "$!"
```

Before: 
Screenshot from 2021-01-21 15-58-49

After:
Screenshot from 2021-01-21 16-01-52

(Yes, there's a typo in the service name, but I don't want to redo the test :disappointed:)

File: lib/base/process.cpp [hunk 3]
```diff
@@ -1029,15 +1032,42 @@ bool Process::DoEvents()
...
+				m_SentSigterm = true;
+			}
+		}
+#endif /* _WIN32 */
+
+		auto timeout (GetNextTimeout());
+		auto deadline (m_Result.ExecutionStart + timeout);
 
-		if (timeout < Utility::GetTime()) {
+		if (deadline < now) {
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/base/process.cpp:1064]: "without more ifdefs"... (laughs)... errm... no. :)

**Pull request merged**